Studio-Lovelies/GG-JointJustice-Unity
# (Re)Integrate automated PlayMode and EditMode tests to build pipeline

## Summary
### Integrate automated PlayMode and EditMode tests to build pipeline
Merging https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/30 caused automated GitHub Actions-based builds hanging and being killed by a 2 hour timeout, resulting in being unable to create new releases. This PR addresses the underlying issue (caching of Git LFS and the `Library`-folder) and reinstates running PlayMode and EditMode tests as part of our GitHub Actions workflow.   
It also addresses a rare side-issue, where [game code compilation errors or "no tests" incorrectly reported as "successful test runs"](https://github.com/game-ci/unity-test-runner/pull/137) by [upgrading the test runner dependency](https://github.com/game-ci/unity-test-runner/releases/tag/v2.0-alpha-6#:~:text=%23143%20Use%20exit%20code%20to%20fail%20when%20code%20compile%20fails).  
Additionally, tests and builds are now ran in parallel. Due to this, this change only adds around 30-60 seconds to our existing workflow, instead of 5 minutes.  
Releases (on GitHub Releases or itch.io) are still only created if the game build process AND tests succeed.

### Added "Running tests locally" and "Contributing" chapters to the `README.MD`-file
Pretty self-explanatory. The `README.MD`-file now contains step-by-step instructions to run automated tests in-editor + links to the code style guide inside our Google Drive.

## Before/after screenshots and/or animated gif
### Before
![image](https://user-images.githubusercontent.com/1689033/132991827-0e672b08-d516-4585-880c-dfa60f3ebc2d.png)
### After
![image](https://user-images.githubusercontent.com/1689033/132991849-ccb139d7-2515-4415-817c-2d876e0300ce.png)
(I'm really not the biggest fan of GitHub's way of visualizing pipelines. 👀)

## Testing instructions
1. Manually [queue a build for this branch](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/actions/workflows/build.yml)
2. Wait for the run to start and open it's sub-page ([example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/actions/runs/1226709572))
2. Wait ~10 minutes
3. Check the `Test Results` tab on the left and see all present tests marked as successful ([example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/runs/3579891372?check_suite_focus=true))
  4.1. Expand the `.xml`-files at the bottom and see the runtime for individual test cases
4. Check the `Summary` tab on the left, scroll all the way down and see builds for all supported platforms + Test Results being available as downloadable artifacts ([example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/actions/runs/1226709572))
5. (Optional: Branch off from this branch, deliberately break a test case, repeat the steps outlined above and find a call stack to your change inside the `playmode.log` of the `Test results` artifact.)

## Additional information

- `[ ]` Changes UI
- `[x]` Introduces new feature
  - Execution of automated test suites inside build workflow
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  - https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/30
  - https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/51
